src/preferences.cpp

```diff
@@ -0,0 +1,546 @@
+ */
+void Preferences::change_mouse_sensitivity(int value)
+{
+	mouse_sensivity = value;
+}
+
+/**
+ * @brief Preferences::change_key_inputs
+ * this changes the keyboard maps and saves those changes in the config file.
+ * this function is executed if the user presses ok button.
+ */
+
+void Preferences::change_key_inputs()
+{
+    //change and save key maps
+    GAMEPAD_BUTTONS[GamepadButtons::GamepadButtons_X] = this->X;
+    save_setting(keymaps[setting_keys::keys::X], this->X);
+    /*------------------------------------------------------------*/
+    GAMEPAD_BUTTONS[GamepadButtons::GamepadButtons_Y] = this->Y;
+    save_setting(keymaps[setting_keys::keys::Y], this->Y);
+    /*------------------------------------------------------------*/
+    GAMEPAD_BUTTONS[GamepadButtons::GamepadButtons_A] = this->A;
+    save_setting(keymaps[setting_keys::keys::A], this->A);
+    /*------------------------------------------------------------*/
+    GAMEPAD_BUTTONS[GamepadButtons::GamepadButtons_B] = this->B;
+    save_setting(keymaps[setting_keys::keys::B], this->B);
+    /*------------------------------------------------------------*/
+    GAMEPAD_BUTTONS[GamepadButtons::GamepadButtons_LeftThumbstick] = this->LT;
+    save_setting(keymaps[setting_keys::keys::LT], this->LT);
+    /*------------------------------------------------------------*/
+    GAMEPAD_BUTTONS[GamepadButtons::GamepadButtons_RightThumbstick] = this->RT;
+    save_setting(keymaps[setting_keys::keys::RT], this->RT);
+    /*------------------------------------------------------------*/
+    GAMEPAD_BUTTONS[GamepadButtons::GamepadButtons_DPadDown] = this->DDOWN;
+    save_setting(keymaps[setting_keys::keys::DPADDOWN], this->DDOWN);
+    /*------------------------------------------------------------*/
+    GAMEPAD_BUTTONS[GamepadButtons::GamepadButtons_DPadUp] = this->DUP;
+    save_setting(keymaps[setting_keys::keys::DPADUP], this->DUP);
+    /*------------------------------------------------------------*/
+    GAMEPAD_BUTTONS[GamepadButtons::GamepadButtons_DPadRight] = this->DRIGHT;
+    save_setting(keymaps[setting_keys::keys::DPADRIGHT], this->DRIGHT);
+    /*------------------------------------------------------------*/
+    GAMEPAD_BUTTONS[GamepadButtons::GamepadButtons_DPadLeft] = this->DLEFT;
+    save_setting(keymaps[setting_keys::keys::DPADLEFT], this->DLEFT);
+}
+
+/**
+ * @brief Preferences::get_scan_code
+ * gets the name of the of corrosponding key or virutal key code.
+ * @param vk
+ * the virtual key code of the key you want to get.
+ * @param a
+ * the buffer to store the name of the key.
+ * @param size
+ * size of the buffer in which the name is stored to ensure memory safety
+ */
+void Preferences::get_scan_code(WORD vk, char* a, int size)
+{
+    char sc = MapVirtualKeyA((UINT)vk, MAPVK_VK_TO_CHAR);
+    if(sc >= '0' && sc <= 'Z') {
+        strncpy(a, "", size);
+        strncpy(a, &sc, 1);
+    }
+    else {
+        strncpy(a, vk_maps[vk], size);
+    }
+}
```

Comment:
@franklin654
There was a bug in the code using `strncpy` that was causing the application to crash.  
Did you fix that?  
Please debug and verify.

See `Preferences::get_scan_code()` in the file src/preferences.cpp.

Reply:
Yes I replaced the strncpy with strncpy_s with buffer size checks just haven't commited yet there are other changes that need to be done in preferences.cpp